src/main/java/com/serenitydojo/ShoppingCart.java

```diff
@@ -0,0 +1,74 @@
+package com.serenitydojo;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ShoppingCart {
+    /*Ash to reviewer: when adding items to shopping cart, should you add catalog item? or fruit?
+    * */
+    Catalog catalog;
+    Map<Fruit, Integer> fruitToQuantityMap;
+
+    Double discountPercentage = 0.10;
+
+
+    public ShoppingCart(Catalog catalog) {
+        this.catalog = catalog;
+        fruitToQuantityMap = new HashMap<>();
+    }
+
+// This method adds item (Fruit) and marks it for Discount if applicable, to apply to total later
+// Ash to reviewer: When adding items to cart, I thought it made sense to add the Fruit directly as opposed to adding a CatalogItem.
+//What are your thoughts?
+    public void addItem(Fruit fruit, Integer quantityInKilos) {
+        fruitToQuantityMap.put(fruit, quantityInKilos);
+        if(catalog.checkEligibilityForDiscount(quantityInKilos)) {
+            catalog.markFruitEligibleForDiscount(fruit);
+        }
+    }
+
+    public Map<Fruit, Integer> showItemsInCart() {
+        System.out.println("Items in shopping cart: " + fruitToQuantityMap);
+        return fruitToQuantityMap;
+    }
+
+    public Double getTotalPriceOfCartBeforeDiscounts() throws FruitUnavailableException {
+        Double totalPrice = 0.0;
+        Double totalPriceAfterDiscount = 0.0;
+        for (Map.Entry<Fruit, Integer> entry : fruitToQuantityMap.entrySet()) {
+            Fruit fruit = entry.getKey();
+            Integer quantityInKilos = entry.getValue();
+            totalPrice = totalPrice + catalog.calculateFruitPriceForQuantity(fruit, quantityInKilos);
+        }
+
+        System.out.println("Total price before discount from method getTotalPriceOfCartBeforeDiscounts : " + totalPrice);
+        return totalPrice;
+    }
+
+    public Double getTotalPriceOfCartAfterDiscounts() throws FruitUnavailableException {
+        Double totalPriceAfterDiscount = 0.0;
+        Double totalPriceBeforeDiscount = getTotalPriceOfCartBeforeDiscounts();
+
+        //E.g: 18 = 20 - (20*0.1)
+        totalPriceAfterDiscount =  totalPriceBeforeDiscount - getDiscountOffer(totalPriceBeforeDiscount);
+        System.out.println("Total price after discount from method getTotalPriceOfCartAfterDiscounts : " + totalPriceAfterDiscount);
+
+        return totalPriceAfterDiscount;
+    }
+
+    public Double getDiscountOffer(Double totalPriceBeforeDiscounts){
+        Double discountAmount = 0.0;
+        //if there's one or more fruits in the eligible List for discount, then apply 10% discount on the total cart price
+        if(catalog.fruitsEligibleForDiscount.size() >= 1) {
```

Comment:
the whole discount stuff seems more complicated than need be....
like.. instead of `catalog.fruitsEligibleForDiscount` i would have that be part of the `cart`.. seems more logical to me that you keep track of discounts in the cart and not in the catalog. then.. if i recall correctly, the requirement is "if any item is > 5kg, you get 10% discount". it's not clear if that's on the whole cart or just on that cart item...

if "on the whole cart" you could have in the `cart` class a `boolean discounted = false` and when you add an item >5k you could set `discounted = true`.... or instead of `catalog.fruitsEligibleForDiscount.size() >= 1` I'd do `fruitToQuantity.values().stream().anyMatch(value -> value >=5)`

Reply:
A million thanks for pinpointing this! All along the voice in my head said that the solution couldn't need this level of complexity. That said I struggled with coming up with a simpler design.  Feeling a little defeated, I went ahead with my overly complicated code so I wouldn't feel as stuck.

Reply:
:) things can always be simpler. and - at least to me - you need to have some distance to your work to see that. it's good that you just went it head with that though.. because now, you have test coverage and something that works as intended. when you realize after 2 weeks that things could be much simpler, you can rip apart a lot of your code internals and your tests are still valid. you can safely change implementation and have the feedback "all green, nothing broke, simpler works as well as the code before".